src/mlpack/methods/reinforcement_learning/q_learning_impl.hpp

```diff
@@ -121,8 +174,20 @@ double QLearning<
   arma::colvec sampledRewards;
   arma::mat sampledNextStates;
   arma::icolvec isTerminal;
-  replayMethod.Sample(sampledStates, sampledActions, sampledRewards,
-      sampledNextStates, isTerminal);
+  arma::ucolvec sampledIndices;
+  arma::rowvec weights;
+
+  if (!prioritized_replay)
+  {
+    replayMethod.Sample(sampledStates, sampledActions, sampledRewards,
+        sampledNextStates, isTerminal);
+  }
+  else
+  {
+    prioritizedReplayMethod.Sample(sampledStates, sampledActions,
```

Comment:
What I had in mind was to remove the ``prioritizedReplayMethod`` template parameter and use the ``policy`` parameter instead; since the prioritized replay sample method takes the sampled indicies and weights, we would have to pass that to the other replay method as well. So,

https://github.com/mlpack/mlpack/blob/85e00107fe3e2bb8d12f331a725cdfe11acb290b/src/mlpack/methods/reinforcement_learning/replay/random_replay.hpp#L111-L115

becomes:

```c++
void Sample(arma::mat& sampledStates,
              arma::icolvec& sampledActions,
              arma::colvec& sampledRewards,
              arma::mat& sampledNextStates,
              arma::icolvec& isTerminal,
              arma::ucolvec& /*sampledIndices */,
              arma::rowvec& /*weights*/)
```

we just pass the two extra parameter, but we don't use them.

Reply:
sorry for late reply, I am not feeling well recently.  I will take a look, thanks for the review.

Reply:
No worries, hope you are feeling better soon.

Reply:
> No worries, hope you are feeling better soon.

sorry for being away for so long, I am back and will take a look.

Reply:
Great, thanks!

Reply:
> What I had in mind was to remove the `prioritizedReplayMethod` template parameter and use the `policy` parameter instead; since the prioritized replay sample method takes the sampled indicies and weights, we would have to pass that to the other replay method as well. So,
> 
> https://github.com/mlpack/mlpack/blob/85e00107fe3e2bb8d12f331a725cdfe11acb290b/src/mlpack/methods/reinforcement_learning/replay/random_replay.hpp#L111-L115
> 
> becomes:
> 
> ```c++
> void Sample(arma::mat& sampledStates,
>               arma::icolvec& sampledActions,
>               arma::colvec& sampledRewards,
>               arma::mat& sampledNextStates,
>               arma::icolvec& isTerminal,
>               arma::ucolvec& /*sampledIndices */,
>               arma::rowvec& /*weights*/)
> ```
> 
> we just pass the two extra parameter, but we don't use them.

Hi, @zoq, please let me clarify to make sure I didn't miss anything.  
You mean remove the 'prioritizedReplayMethod' parameter 
```
arma::ucolvec& sampledIndices,
arma::rowvec& weights
```

But do you mean move those two parameters into `policy` class? I found that we can move them into `prioritizedReplayMethod`, then we can access them as a class parameter. What do you think of it?

Reply:
It's look like a discussion from way back. I doubt it may not even be active now.
@zoq Sorry, but I can't seem to get, what do you mean by removing `prioritizedReplayMethod` template parameter here? If this discussion is still applicable, can you please explain it a bit?

Reply:
What I meant was to switch from:

```c++
template <
  typename EnvironmentType,
  typename NetworkType,
  typename UpdaterType,
  typename PolicyType,
  typename ReplayType = RandomReplay<EnvironmentType>,
  typename PrioritizedReplayType = PrioritizedReplay<EnvironmentType>
>
class QLearning
```

to what we already have

```c++
template <
  typename EnvironmentType,
  typename NetworkType,
  typename UpdaterType,
  typename PolicyType,
  typename ReplayType = RandomReplay<EnvironmentType>>
>
class QLearning
```

so instead of adding another template parameter (``PrioritizedReplayType``) why not reuse ``ReplayType``, which we could use like:


```c++
// Set up the network.
FFN<MeanSquaredError<>, GaussianInitialization> model(MeanSquaredError<>(),
  GaussianInitialization(0, 0.001));
model.Add<Linear<>>(4, 128);
model.Add<ReLULayer<>>();
model.Add<Linear<>>(128, 128);
model.Add<ReLULayer<>>();
model.Add<Linear<>>(128, 2);

// Set up the policy and replay method.
GreedyPolicy<CartPole> policy(1.0, 1000, 0.1, 0.99);
PrioritizedReplay<CartPole> replayMethod(10, 10000);	// Was RandomReplay<CartPole> replayMethod(10, 10000);

TrainingConfig config;
config.StepSize() = 0.01;
config.Discount() = 0.9;
config.TargetNetworkSyncInterval() = 100;
config.ExplorationSteps() = 100;
config.DoubleQLearning() = false;
config.StepLimit() = 200;

// Set up DQN agent.
QLearning<CartPole, decltype(model), AdamUpdate, decltype(policy)>
  agent(std::move(config), std::move(model), std::move(policy),
  std::move(replayMethod));
```

in the example above I just switch the type of the replay method from ``RandomReplay`` to ``PrioritizedReplay``.

The ``PrioritizedReplay`` class expects some additional parameters, but in some cases the interface is already the same, e.g.:


```c++
// Store the transition for replay.
if (prioritized_replay)
{
prioritizedReplayMethod.Store(state, action, reward,
    nextState, environment.IsTerminal(nextState));
}
else
{
replayMethod.Store(state, action, reward,
    nextState, environment.IsTerminal(nextState));
}
```

can just be:

```c++
replayMethod.Store(state, action, reward,
    nextState, environment.IsTerminal(nextState));
```

the same applies to the ``Sample`` and ``Update``. So to sum this up, In don't see a reason to add another template parameter for Prioritized Experience Replay. Does that sound reasonable?

Reply:
Thanks for thorough explanation. Right, that totally make sense. It would be best if we can preserve the template structure and avoid adding extra if else conditions. And your idea completely make up for it.
@robotcator please do have a look at this, and if you like you can make the required changes here itself.

Reply:
> What I meant was to switch from:
> 
> ```c++
> template <
>   typename EnvironmentType,
>   typename NetworkType,
>   typename UpdaterType,
>   typename PolicyType,
>   typename ReplayType = RandomReplay<EnvironmentType>,
>   typename PrioritizedReplayType = PrioritizedReplay<EnvironmentType>
> >
> class QLearning
> ```
> 
> to what we already have
> 
> ```c++
> template <
>   typename EnvironmentType,
>   typename NetworkType,
>   typename UpdaterType,
>   typename PolicyType,
>   typename ReplayType = RandomReplay<EnvironmentType>>
> >
> class QLearning
> ```
> 
> so instead of adding another template parameter (`PrioritizedReplayType`) why not reuse `ReplayType`, which we could use like:
> 
> ```c++
> // Set up the network.
> FFN<MeanSquaredError<>, GaussianInitialization> model(MeanSquaredError<>(),
>   GaussianInitialization(0, 0.001));
> model.Add<Linear<>>(4, 128);
> model.Add<ReLULayer<>>();
> model.Add<Linear<>>(128, 128);
> model.Add<ReLULayer<>>();
> model.Add<Linear<>>(128, 2);
> 
> // Set up the policy and replay method.
> GreedyPolicy<CartPole> policy(1.0, 1000, 0.1, 0.99);
> PrioritizedReplay<CartPole> replayMethod(10, 10000);	// Was RandomReplay<CartPole> replayMethod(10, 10000);
> 
> TrainingConfig config;
> config.StepSize() = 0.01;
> config.Discount() = 0.9;
> config.TargetNetworkSyncInterval() = 100;
> config.ExplorationSteps() = 100;
> config.DoubleQLearning() = false;
> config.StepLimit() = 200;
> 
> // Set up DQN agent.
> QLearning<CartPole, decltype(model), AdamUpdate, decltype(policy)>
>   agent(std::move(config), std::move(model), std::move(policy),
>   std::move(replayMethod));
> ```
> 
> in the example above I just switch the type of the replay method from `RandomReplay` to `PrioritizedReplay`.
> 
> The `PrioritizedReplay` class expects some additional parameters, but in some cases the interface is already the same, e.g.:
> 
> ```c++
> // Store the transition for replay.
> if (prioritized_replay)
> {
> prioritizedReplayMethod.Store(state, action, reward,
>     nextState, environment.IsTerminal(nextState));
> }
> else
> {
> replayMethod.Store(state, action, reward,
>     nextState, environment.IsTerminal(nextState));
> }
> ```
> 
> can just be:
> 
> ```c++
> replayMethod.Store(state, action, reward,
>     nextState, environment.IsTerminal(nextState));
> ```
> 
> the same applies to the `Sample` and `Update`. So to sum this up, In don't see a reason to add another template parameter for Prioritized Experience Replay. Does that sound reasonable?

But I have a question about that,  how do we know which `replay buffer` passed into the q_learning class?  If we don't use inheritance, how to deal with the situation when we have two constructors?

```
QLearning(TrainingConfig config,
            NetworkType network,
            PolicyType policy,
            ReplayType replayMethod,
            UpdaterType updater = UpdaterType(),
            EnvironmentType environment = EnvironmentType());

  QLearning(TrainingConfig config,
            NetworkType network,
            PolicyType policy,
            PrioritizedReplayType prioritizedReplayMethod,
            UpdaterType updater = UpdaterType(),
            EnvironmentType environment = EnvironmentType());
```

Reply:
> But I have a question about that, how do we know which replay buffer passed into the q_learning class? 

Sorry, I don't seem to understand the question. Isn't the type of replay you pass to the Qlearning agent will be the one, it will be using. Or am I missing something here?

> If we don't use inheritance, how to deal with the situation when we have two constructors?

We actually don't need to have two separate constructors, if we reuse the ReplayType template parameter instead of having a separate template parameter of prioritized replay. Does it clear things up?

Reply:
Right, in the example above, ``PrioritizedReplayType`` would  just be ``ReplayType`, and ass @manish7294 already pointed out this makes the second constructor obsolete.

Reply:
> Right, in the example above, `PrioritizedReplayType` would just be ``ReplayType`, and ass @manish7294 already pointed out this makes the second constructor obsolete.

Sorry, it seems that I missing something. I do want to reuse the `ReplayType`, but I didn't realize that we can reuse the template parameter.

Now, things are all cleared. But if we use `prioritizedReplayMethod`, we should use the Q-learning class like this:
```
  QLearning<CartPole, decltype(model), AdamUpdate, decltype(policy),
      decltype(replayMethod)>
    agent(std::move(config), std::move(model), std::move(policy),
          std::move(replayMethod));
```
because the default replaymethod is `RandomReplay`.

Reply:
True, we have to pass the correct type for the replay method.